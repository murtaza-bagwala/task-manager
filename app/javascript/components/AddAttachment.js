import React, { useState } from 'react';

import { connect } from 'react-redux';
import Button from '@material-ui/core/Button';
import { saveFile } from '../redux/attachmentActions';

import { ALLOWED_EXTENSIONS } from '../constants';

export const AddAttachment = ({ taskId, saveFile }) => {
  const [error, setError] = useState('');

  const onChange = async (file) => {
    if (file && ALLOWED_EXTENSIONS.includes(file.name.split('.').pop())) {
      const token = sessionStorage.getItem('token');
      saveFile(token, taskId, file);
      setError('');
    } else {
      setError('Extension not allowed');
    }
  };

  return (
    <label htmlFor="upload-photo">
      <input
        style={{ display: 'none' }}
        id="upload-photo"
        name="upload-photo"
        type="file"
        onChange={(e) => onChange(e.target.files[0])}
      />

      <Button color="primary" variant="contained" component="span">
        Upload File
      </Button>
      <span style={{ color: 'red' }}>{error}</span>
    </label>
  );
};

export default connect(null, { saveFile })(AddAttachment);
