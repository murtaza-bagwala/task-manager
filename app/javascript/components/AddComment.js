import React, { useState } from 'react';

import { connect } from 'react-redux';
import TextField from '@material-ui/core/TextField';
import { saveComment } from '../redux/commentActions';

export const AddComment = ({ saveComment, taskId }) => {
  const [state, setState] = useState({});

  const updateInput = (input) => {
    setState({ input });
  };

  const handleAddComment = (e) => {
    if (e.keyCode === 13 && state.input && state.input.length) {
      const token = sessionStorage.getItem('token');
      saveComment(token, taskId, state.input);
      setState({ input: '' });
    }
  };

  return (
    <TextField
      id="standard-full-width"
      style={{ margin: 8 }}
      fullWidth
      margin="normal"
      onChange={(e) => updateInput(e.target.value)}
      value={state.input}
      helperText={!state.input ? 'Comment cannot be empty!' : ''}
      error={!state.input ? 'Comment cannot be empty!' : ''}
      onKeyDown={handleAddComment}
    />
  );
};

export default connect(null, { saveComment })(AddComment);
