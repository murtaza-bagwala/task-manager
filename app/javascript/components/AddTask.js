import React, { useState } from 'react';

import { connect } from 'react-redux';
import TextField from '@material-ui/core/TextField';
import { saveTask } from '../redux/actions';

export const AddTask = ({ saveTask }) => {
  const [state, setState] = useState({});

  const updateInput = (input) => {
    setState({ input });
  };

  const handleAddTask = (e) => {
    if (e.keyCode === 13 && (state.input && state.input.length)) {
      const token = sessionStorage.getItem('token');
      saveTask(token, {
        content: state.input,
        deadline: '',
      });
      setState({ input: '' });
    }
  };

  return (
    <TextField
      id="standard-full-width"
      style={{ margin: 8 }}
      placeholder="What needs to be done?"
      fullWidth
      margin="normal"
      onChange={(e) => updateInput(e.target.value)}
      value={state.input}
      onKeyDown={handleAddTask}
      helperText={!state.input ? 'Task description cannot be empty!' : ''}
      error={!state.input ? 'Task description cannot be empty!' : ''}
    />
  );
};

export default connect(null, { saveTask })(AddTask);
