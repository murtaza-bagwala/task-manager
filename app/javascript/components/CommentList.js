import React from 'react';
import { connect } from 'react-redux';
import { makeStyles } from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import { getCommentsForTask } from '../redux/selectors';

const useStyles = makeStyles((theme) => ({
  text: {
    padding: theme.spacing(2, 2, 0),
  },
  textField: {},
  paper: {
    paddingBottom: 50,
    maxHeight: 200,
    overflow: 'auto',
  },
  list: {
    marginBottom: theme.spacing(2),
  },
  subheader: {
    backgroundColor: theme.palette.background.paper,
  },
  fabButton: {
    position: 'absolute',
    zIndex: 1,
    top: -30,
    left: 0,
    right: 0,
    margin: '0 auto',
  },
}));

export const CommentList = ({ comments = [] }) => {
  const classes = useStyles();

  return (
    <Paper square className={classes.paper}>
      <List className={classes.list}>
        {comments.map((comment, index) => (
          <ListItem>
            <ListItemText primary={comment.comment} />
          </ListItem>
        ))}
      </List>
    </Paper>
  );
};

const mapStateToProps = (state) => {
  const { selectedTask } = state.tasks;
  debugger;
  const comments = getCommentsForTask(state, selectedTask.comments);
  debugger;
  return { comments };
};

export default connect(mapStateToProps, null)(CommentList);
