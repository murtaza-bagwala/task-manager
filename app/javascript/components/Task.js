import React from 'react';
import { connect } from 'react-redux';
import cx from 'classnames';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import EditRoundedIcon from '@material-ui/icons/EditRounded';
import DeleteIcon from '@material-ui/icons/Delete';
import { destroyTask, setSelectedTask } from '../redux/actions';

export const Task = ({
  task, toggleTask, destroyTask, setSelectedTask,
}) => {
  const handleDelete = () => {
    const token = sessionStorage.getItem('token');
    destroyTask(token, task.id);
  };

  return (
    <React.Fragment key={task.id}>
      <ListItem
        button
      >
        <ListItemText primary={task.content} />
        <EditRoundedIcon onClick={() => setSelectedTask(task)} />
        <DeleteIcon onClick={handleDelete} />
      </ListItem>
    </React.Fragment>
  );
};

export default connect(null, { destroyTask, setSelectedTask })(Task);
