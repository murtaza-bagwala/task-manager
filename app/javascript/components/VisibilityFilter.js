import React from 'react';
import cx from 'classnames';
import { connect } from 'react-redux';
import Paper from '@material-ui/core/Paper';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';
import { VISIBILITY_FILTERS } from '../constants';
import { setFilter } from '../redux/actions';

export const VisibilityFilters = ({ activeFilter, setFilter }) => (
  <Paper square>
    <Tabs indicatorColor="primary" textColor="primary">
      {Object.keys(VISIBILITY_FILTERS).map((filterKey) => {
        const currentFilter = VISIBILITY_FILTERS[filterKey];
        return (
          <Tab
            label={filterKey}
            key={`visibility-filter-${currentFilter}`}
            onClick={() => {
              setFilter(currentFilter);
            }}
          />
        );
      })}
    </Tabs>
  </Paper>
);

const mapStateToProps = (state) => ({ activeFilter: state.visibilityFilter });
// export default VisibilityFilters;
export default connect(mapStateToProps, { setFilter })(VisibilityFilters);
