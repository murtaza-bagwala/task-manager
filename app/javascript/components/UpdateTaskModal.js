import { makeStyles } from '@material-ui/core/styles';
import Modal from '@material-ui/core/Modal';
import React, { useState } from 'react';

import TextField from '@material-ui/core/TextField';
import Checkbox from '@material-ui/core/Checkbox';
import CancelIcon from '@material-ui/icons/Cancel';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import { connect } from 'react-redux';
import CommentList from './CommentList';
import AddComment from './AddComment';
import AddAttachment from './AddAttachment';

import { getAttachmentsForTask } from '../redux/selectors';

import { modifyTask, setSelectedTask } from '../redux/actions';

import { ATTACHMENT_HOST } from '../constants';

const isEmpty = (object) => !object || Object.keys(object).length === 0;

const useStyles = makeStyles((theme) => ({
  paper: {
    position: 'absolute',
    width: 400,
    backgroundColor: theme.palette.background.paper,
    border: '2px solid #000',
    boxShadow: theme.shadows[5],
    padding: theme.spacing(2, 4, 3),
  },
}));

export const UpdateTaskModal = ({
  attachments,
  selectedTask,
  modifyTask,
  setSelectedTask,
}) => {
  const classes = useStyles();

  const [content, setContent] = useState({});
  const [error, setError] = useState(false);

  const updateContent = (taskContent) => {
    setContent({ taskContent });
  };

  const toggleCheckbox = (completed) => {
    const token = sessionStorage.getItem('token');
    modifyTask(token, {
      ...selectedTask,
      completed,
    });
  };

  const updateDeadline = (deadline) => {
    const currentDate = new Date();
    const taskDeadline = new Date(deadline);
    if (taskDeadline > currentDate) {
      const token = sessionStorage.getItem('token');
      modifyTask(token, {
        ...selectedTask,
        deadline,
      });
      setError(false);
    } else {
      setError(true);
    }
  };

  const handleEditContent = (keyCode) => {
    if (keyCode === 13 && content.taskContent && content.taskContent.length) {
      const { taskContent } = content;
      const token = sessionStorage.getItem('token');
      modifyTask(token, {
        ...selectedTask,
        content: taskContent,
      });
    }
  };

  const handleClose = () => {
    setSelectedTask({});
    setContent({});
  };

  if (isEmpty(selectedTask)) {
    return null;
  }

  return (
    <div>
      <Modal
        open
        onClose={handleClose}
        style={{
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'center',
        }}
      >
        <div className={classes.paper}>
          <div style={{ 'padding-left': '330px' }}>
            <CancelIcon
              variant="contained"
              color="primary"
              onClick={handleClose}
            />
          </div>
          <TextField
            id="standard-full-width"
            style={{ margin: 8 }}
            placeholder="What needs to be done?"
            fullWidth
            margin="normal"
            value={
              !isEmpty(content) ? content.taskContent : selectedTask.content
            }
            onChange={(e) => updateContent(e.target.value)}
            onClose={handleClose}
            onKeyDown={(e) => handleEditContent(e.keyCode)}
          />
          <div style={{ 'padding-bottom': '10px', 'padding-up': '10px' }}>
            <FormControlLabel
              control={(
                <Checkbox
                  checked={selectedTask.completed}
                  onChange={(e) => toggleCheckbox(e.target.checked)}
                  name="checkedB"
                  color="primary"
                />
              )}
              label="Completed"
            />
          </div>
          <div style={{ 'padding-bottom': '10px', 'padding-up': '10px' }}>
            <TextField
              id="datetime-local"
              label="Set Deadline"
              type="datetime-local"
              defaultValue={
                selectedTask.deadline
                  ? selectedTask.deadline.replace(/.\d+Z$/g, '')
                  : '2021-06-01T10:30'
              }
              className={classes.textField}
              onChange={(e) => updateDeadline(e.target.value)}
              InputLabelProps={{
                shrink: true,
              }}
              helperText={error ? 'Deadline should be greater than the currentDate' : ''}
              error={error ? 'Deadline should be greater than the currentDate' : ''}
            />
          </div>
          <div style={{ 'padding-bottom': '10px', 'padding-up': '10px' }}>
            <AddComment taskId={selectedTask.id} />
            <CommentList />
          </div>
          <div style={{ 'padding-bottom': '10px', 'padding-up': '10px' }}>
            <AddAttachment taskId={selectedTask.id} />
          </div>
          <div style={{ 'padding-bottom': '10px', 'padding-up': '10px' }}>
            {attachments.map((attachment) => (
              <a href={`${ATTACHMENT_HOST}/${attachment.url}`} target="_blank" rel="noreferrer">{attachment.name}</a>
            ))}
          </div>
        </div>
      </Modal>
    </div>
  );
};

const mapStateToProps = (state) => {
  const { selectedTask } = state.tasks;
  let attachments = [];
  if (selectedTask.attachments) {
    attachments = getAttachmentsForTask(state, selectedTask.attachments);
  }
  return { selectedTask, attachments };
};

export default connect(mapStateToProps, { modifyTask, setSelectedTask })(
  UpdateTaskModal,
);
