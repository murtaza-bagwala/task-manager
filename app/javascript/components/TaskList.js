import React, { useEffect } from 'react';

import { connect } from 'react-redux';
import { makeStyles } from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import List from '@material-ui/core/List';
import { getTasksByVisibilityFilter } from '../redux/selectors';
import Task from './Task';
import UpdateTaskModal from './UpdateTaskModal';
import { fetchTasks } from '../redux/actions';

const useStyles = makeStyles((theme) => ({
  text: {
    padding: theme.spacing(2, 2, 0),
  },
  textField: {},
  paper: {
    paddingBottom: 50,
    maxHeight: 400,
    overflow: 'auto',
  },
  list: {
    marginBottom: theme.spacing(2),
  },
  subheader: {
    backgroundColor: theme.palette.background.paper,
  },
  fabButton: {
    position: 'absolute',
    zIndex: 1,
    top: -30,
    left: 0,
    right: 0,
    margin: '0 auto',
  },
}));

export const TaskList = ({ tasks, fetchTasks }) => {
  const classes = useStyles();

  useEffect(() => {
    const token = sessionStorage.getItem('token');
    fetchTasks(token);
  }, []);

  return (
    <Paper square className={classes.paper}>
      <UpdateTaskModal />
      <List className={classes.list}>
        {tasks.map((task, index) => (
          <Task task={task} />
        ))}
      </List>
    </Paper>
  );
};

const mapStateToProps = (state) => {
  const { visibilityFilter } = state;
  const tasks = getTasksByVisibilityFilter(state, visibilityFilter);
  return { tasks };
};

export default connect(mapStateToProps, { fetchTasks })(TaskList);
