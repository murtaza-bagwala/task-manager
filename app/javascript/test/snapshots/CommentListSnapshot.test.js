import React from 'react';
import renderer from 'react-test-renderer';
import { CommentList } from '../../components/CommentList';

it('renders CommentList Successfully', () => {
  const tree = renderer.create(<CommentList />);
  expect(tree).toMatchSnapshot();
});
