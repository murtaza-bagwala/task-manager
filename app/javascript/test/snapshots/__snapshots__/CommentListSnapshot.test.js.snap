// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders CommentList Successfully 1`] = `
<div
  className="MuiPaper-root makeStyles-paper-3 MuiPaper-elevation1"
>
  <ul
    className="MuiList-root makeStyles-list-4 MuiList-padding"
  />
</div>
`;
